sdk/core/azure-core/inc/uuid.hpp

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <cstdlib> // for abort
+#include <new> // for placement new
+#include <string>
+#include <type_traits>
+#include <utility> // for swap and move
+
+namespace Azure { namespace Core {
+
+  class UUID {
+
+  private:
+    static const size_t UUIDSize = 16;
+
+    unsigned char m_uuid[UUIDSize];
+    // The UUID reserved variants.
+    static const unsigned char ReservedNCS = 0x80;
+    static const unsigned char ReservedRFC4122 = 0x40;
+    static const unsigned char ReservedMicrosoft = 0x20;
+    static const unsigned char ReservedFuture = 0x00;
+
+  public:
+    UUID()
+    {
+      std::random_device rd;
+      std::mt19937 gen(rd());
+      for (int i = 0; i < UUIDSize; i++)
+        m_uuid[i] = (unsigned char)gen();
+
+      //SetVariant to ReservedRFC4122
+      m_uuid[8] = (m_uuid[8] | ReservedRFC4122) & 0x7F;
+
+      const unsigned char version = 4;
+
+      m_uuid[6] = (m_uuid[6] & 0xF) | (version << 4);
+    }
+
+    std::string GetUUIDString() {
+      char s[36];
+
+      sprintf(&s[0], "%x%x%x%x-%x%x-%x%x-%x%x-%x%x%x%x%x%x",
```

Comment:
%x does not work as you expect for the values 0x00 - 0x0F

Reply:
the format string needs to be `"%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x"`

Reply:
```
char formatBuffer[sizeof("1A2B3C4D-1A2B-1A2B-1A2B-1A2B3C4D5E6F")] = {0};
  constexpr auto formatString
      = "%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x";

  std::sprintf(
      formatBuffer,
      formatString,
      m_uuid[0],
      m_uuid[1],
      m_uuid[2],
      m_uuid[3],
      m_uuid[4],
      m_uuid[5],
      m_uuid[6],
      m_uuid[7],
      m_uuid[8],
      m_uuid[9],
      m_uuid[10],
      m_uuid[11],
      m_uuid[12],
      m_uuid[13],
      m_uuid[14],
      m_uuid[15]);
```
(and it is better to Ctrl+K-D (autoformat), because it will eventially happen when someone modifies it. And I don't think that m_uuids one under another look THAT bad, I even see some benefit in indices being one under another; plus you don't need to update ends of lines if we change places where we put dashes)

Reply:
Another, and maybe better in terms of maintenance option would be to use `std::ostringstream`.
You can see the example how I do it in `UrlEncode()` in `credentials.cpp`.
I know that string formatting is faster but I don't think this will ever be a bottleneck. But you don't have to worry about the buffer size.
```
#include <iomanip>
#include <sstream>
...
std::ostringstream s;
s<< std::hex;
s<< std::setw(2) << m_uuid[0] ...
...
return s.str();
```